src/datastore/pg-write-store.ts

```diff
@@ -746,14 +759,35 @@ export class PgWriteStore extends PgStore {
     logger.info('Updated block zero boot data', tablesUpdates);
   }
 
-  async updatePoxSyntheticEvents(
-    sql: PgSqlClient,
-    tx: DbTx,
-    poxTable: PoxSyntheticEventTable,
-    events: DbPoxSyntheticEvent[]
-  ) {
-    for (const batch of batchIterate(events, INSERT_BATCH_SIZE)) {
-      const values = batch.map(event => {
+  async updatePoxSyntheticEvents<
+    T extends PoxSyntheticEventTable,
+    Entry extends { tx: DbTx } & ('pox2_events' extends T
+      ? { pox2Events: DbPoxSyntheticEvent[] }
+      : 'pox3_events' extends T
+      ? { pox3Events: DbPoxSyntheticEvent[] }
+      : 'pox4_events' extends T
+      ? { pox4Events: DbPoxSyntheticEvent[] }
+      : never)
+  >(sql: PgSqlClient, poxTable: T, entries: Entry[]) {
+    const values: PoxSyntheticEventInsertValues[] = [];
+    for (const entry of entries) {
+      let events: DbPoxSyntheticEvent[] | null = null;
+      switch (poxTable) {
+        case 'pox2_events':
+          assert('pox2Events' in entry);
+          events = entry.pox2Events;
+        case 'pox3_events':
+          assert('pox3Events' in entry);
+          events = entry.pox3Events;
+        case 'pox4_events':
+          assert('pox4Events' in entry);
+          events = entry.pox4Events;
+      }
+      if (events == null) {
+        throw new Error(`unknown pox table: ${poxTable}`);
+      }
```

Comment:
Fix switch-case fall-through and simplify types

```suggestion
  async updatePoxSyntheticEvents(
    sql: PgSqlClient,
    poxTable: PoxSyntheticEventTable,
    entries: DataStoreTxEventData[]
  ) {
    const values: PoxSyntheticEventInsertValues[] = [];
    for (const entry of entries) {
      let events: DbPoxSyntheticEvent[];
      switch (poxTable) {
        case 'pox2_events':
          events = entry.pox2Events;
          break;
        case 'pox3_events':
          events = entry.pox3Events;
          break;
        case 'pox4_events':
          events = entry.pox4Events;
          break;
        default:
          throw new Error(`unknown pox table: ${poxTable}`);
      }
```

Reply:
The `entries` does not have to be `DataStoreTxEventData`, using typescript type narrowing is better:

![image](https://github.com/hirosystems/stacks-blockchain-api/assets/432106/7b39853e-948a-40d0-a980-728b3dbf4d3a)

Fixed switch-case fall-through.

Reply:
The readability of `updatePoxSyntheticEvents`'s function signature is questionable, although it is type safer for callers, so I'm okay with it 👍